src/Symfony/Component/HttpClient/Response/ResponseTrait.php

```diff
@@ -0,0 +1,267 @@
+        }
+    }
+
+    private function doDestruct()
+    {
+        // Ensure the request is always sent and that we checked the response code
+
+        if ($this->initializer && null === $this->info['error']) {
+            ($this->initializer)($this);
+            $this->initializer = null;
+        }
+
+        if ($this->checkStatusCode) {
+            $this->checkStatusCode();
+        }
+    }
+
+    /**
+     * @internal
+     */
+    public static function stream(iterable $responses, float $timeout = null): \Generator
+    {
+        $runningResponses = [];
+
+        foreach ($responses as $response) {
+            self::schedule($response, $runningResponses);
+        }
+
+        $lastActivity = microtime(true);
+        $isTimeout = false;
+
+        while (true) {
+            $hasActivity = false;
+            $timeoutMax = 0;
+            $timeoutMin = $timeout ?? INF;
+
+            foreach ($runningResponses as $i => [$multi]) {
+                $responses = &$runningResponses[$i][1];
+                self::perform($multi, $responses);
+
+                foreach ($responses as $j => $response) {
+                    $timeoutMax = $timeout ?? max($timeoutMax, $response->timeout);
+                    $timeoutMin = min($timeoutMin, $response->timeout, 1);
+
+                    if (isset($multi->handlesActivity[$j])) {
+                        // no-op
+                    } elseif (!isset($multi->openHandles[$j])) {
+                        unset($responses[$j]);
+                        continue;
+                    } elseif ($isTimeout) {
+                        $multi->handlesActivity[$j] = [new ErrorChunk($response->offset)];
+                    } else {
+                        continue;
+                    }
+
+                    while ($multi->handlesActivity[$j] ?? false) {
+                        $hasActivity = true;
+                        $isTimeout = false;
+
+                        if (\is_string($chunk = array_shift($multi->handlesActivity[$j]))) {
+                            $chunk = new DataChunk($response->offset, $chunk);
+                            $response->offset += \strlen($chunk);
+                        } elseif (null === $chunk) {
+                            if (null !== $e = $response->info['error'] ?? $multi->handlesActivity[$j][0]) {
+                                $response->info['error'] = $e->getMessage();
+                                $chunk = new ErrorChunk($response->offset, $e);
+                            } else {
+                                $chunk = new LastChunk($response->offset);
+                            }
+                            unset($responses[$j]);
+                            $response->close();
+                        } elseif ($chunk instanceof ErrorChunk) {
+                            unset($responses[$j]);
+                            $isTimeout = true;
+                        }
+
+                        yield $response => $chunk;
```

Comment:
If someone calls `iterator_to_array`, this fails since using an object is invalid for an array key.

How would you feel about yielding a two-tuple like: `yield [$response, $chunk]`. Then in a foreach, you'd do:

```php
foreach ($chunk as [$response, $chunk]) {

}
```

Reply:
If you're calling `stream"()`, that's because you want to iterate over incoming traffic. Putting things in an array would defeat that purpose, e.g. by filling memory. Here, you're supposed to deal with each iteration immediately to benefit from the stream. If you don't need streaming, use e.g. `$response->getContent()` directly instead. With that reasoning, I think it would be counterproductive to optimize for `iterator_ro_array()` here.

Reply:
That is a fair point, just feels unorthodox and could lead to subtle bugs because of it.